lib/color.js

```diff
@@ -0,0 +1,473 @@
+     * @param {Object} meta entity meta object
+     * @return {ColorHSV} corrected color in HSV space
+     */
+    colorCorrected(meta) {
+        return this.hueCorrected(meta).gammaCorrected();
+    }
+}
+
+/**
+ * Converts color temp mireds to Kelvins
+ * @param {number} mireds color temp in mireds
+ * @return {number} color temp in Kelvins
+ */
+function miredsToKelvin(mireds) {
+    return 1000000 / mireds;
+}
+
+/**
+ * Converts color temp in Kelvins to mireds
+ * @param {number} kelvin color temp in Kelvins
+ * @return {number} color temp in mireds
+ */
+function kelvinToMireds(kelvin) {
+    return 1000000 / kelvin;
+}
+
+
+/**
+ * Extracts color information from converters input
+ * @param {Object} value converter value parameter
+ * @return {ColorT} normalized color information
+ */
+function extractColorFromValue(value) {
+    if (value.hasOwnProperty('x') && value.hasOwnProperty('y')) {
+        return {hsv: null, xy: ColorXY.fromObject(value)};
+    } else if (value.hasOwnProperty('r') && value.hasOwnProperty('g') && value.hasOwnProperty('b')) {
+        return {hsv: null, xy: new ColorRGB(value.r / 255, value.g / 255, value.b / 255).toXY()};
+    } else if (value.hasOwnProperty('rgb')) {
+        const [r, g, b] = value.rgb.split(',').map((i) => parseInt(i));
+        return {hsv: null, xy: new ColorRGB(r / 255, g / 255, b / 255).toXY()};
+    } else if (value.hasOwnProperty('hex')) {
+        return {hsv: null, xy: ColorRGB.fromHex(value.hex).toXY()};
+    } else if (typeof value === 'string' && value.startsWith('#')) {
+        return {hsv: null, xy: ColorRGB.fromHex(value).toXY()};
+    } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s') && value.hasOwnProperty('l')) {
+        return {hsv: ColorHSV.fromHSL({hue: value.h, saturation: value.s, lightness: value.l}), xy: null};
+    } else if (value.hasOwnProperty('hsl')) {
+        const [h, s, l] = value.hsl.split(',').map((i) => parseInt(i));
+        return {hsv: ColorHSV.fromHSL({hue: h, saturation: s, lightness: l}), xy: null};
+    } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s') && value.hasOwnProperty('b')) {
+        return {hsv: new ColorHSV(value.h, value.s, value.b), xy: null};
+    } else if (value.hasOwnProperty('hsb')) {
+        const [h, s, b] = value.hsb.split(',').map((i) => parseInt(i));
+        return {hsv: new ColorHSV(h, s, b), xy: null};
+    } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s') && value.hasOwnProperty('v')) {
+        return {hsv: new ColorHSV(value.h, value.s, value.v), xy: null};
+    } else if (value.hasOwnProperty('hsv')) {
+        const [h, s, v] = value.hsv.split(',').map((i) => parseInt(i));
+        return {hsv: new ColorHSV(h, s, v), xy: null};
+    } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s')) {
+        return {hsv: new ColorHSV(value.h, value.s), xy: null};
+    } else if (value.hasOwnProperty('h')) {
+        // XXX: should this be allowed?
```

Comment:
I think this is fine to have partial data, as in ZCL we have commands to interact on partial data.

Reply:
My original concern was about `state.color` updates. Lets say user sets `{"hue": 313, "saturation": 100}` and then changes to `{"saturation": 66}`. Resulting state would be `{..., "color": {"saturation": 66}}`

But this should be fixed in #2441, right?

Reply:
Perhaps the object can have a .isPartial() function to check if we have complete data or not.

Reply:
> My original concern was about `state.color` updates. Lets say user sets `{"hue": 313, "saturation": 100}` and then changes to `{"saturation": 66}`. Resulting state would be `{..., "color": {"saturation": 66}}`
> 

That currently not an issue as z2m will already merge the returned newState with the previous cached state.

> But this should be fixed in #2441, right?

It won't fix that specifically, as mentioned above there is nothing to fix. That PR is to fix the issue where xy, hs, and ct get out of sync.

So when we set XY, after that PR we will update hs and ct to contain values from the XY set. Like ways for setting ct we calculate xy and hs, ...

That PR will also make so we can disable this behavior and not update the state when setting the color, e.g. when using reporting for these attributes on the bulb. Some of them, .e.g Innr will do the conversion in the firmware and report all attributes. So we give the option to take the colors from the bulb too, this is useful when using a remote that can also set colors.

Reply:
> > My original concern was about state.color updates. Lets say user sets {"hue": 313, "saturation": 100} and then changes to {"saturation": 66}. Resulting state would be {..., "color": {"saturation": 66}}
>
>That currently not an issue as z2m will already merge the returned newState with the previous cached state.

Ok, I thought it didn't merge nested objects. If it does then my concern is no longer valid.

> That PR is to fix the issue where xy, hs, and ct get out of sync. (...)

Got it, thanks